### ISSUE TITLE
--notify flag does not work as expected and fires before file has finished uploading

### ISSUE THREAD

**Issue by felipebrahm** [Shopify/themekit#749]
**Describe the bug**
I'm using a custom *livereload* implementation to automatically reload the browser once a file has been uploaded to Shopify.

I'm running themekit with `theme watch --notify=/tmp/theme.update` and a livereload script that watches `theme.update`, then refreshes the page once it detects a change in that file. It isn't working as expected and I still need to add a delay for it work. I added some logging to my app and noticed that themekit is touching the theme.update file *before* finishing uploading the file to Shopify servers.

Here's the output of my app (livereload is watching for all changes for logging purposes but only reloading after changes to *theme.update* are detected):
```
[livereload] change detected: assets/bundle.js
[themekit] processing assets/bundle.js
[livereload] change detected: tmp/theme.update
[livereload] Notifying browser to reload [ 'assets/bundle.js' ]
[themekit] Updated assets/bundle.js
```
As you can see, themekit is updating theme.update before the upload is finished.

**To Reproduce**
1. run `theme watch --notify=/tmp/theme.update`
2. Change a file in your theme.
3. *theme.update* will be touched before the file has finished uploading to Shopify servers.

**Expected behavior**
*theme.update*  should be updated once all files have finished uploading to Shopify servers.

**Comment by felipebrahm** [comment]
I skimmed the source code of themekit and I see that right now the logic for touching the *--notify* file is to "wait until workers have been idled for one second". This doesn't seem right to me. The logic should be to touch it once changed files have finished uploading. Otherwise, there's no point to this flag and I could just watch my local files and add a delay after they change, but the whole point is to be notified when the files are already on Shopify servers.

Or am I misunderstanding the meaning of this *--notify* file?

**Comment by tanema** [comment]
So your skimming is no *quite* right. The events are debounced and buffered before being sent out from the file watcher. The `onEvent` method is blocking so it will not continue in the main `watchFsEvents` until all events are processed. So this means that the notify file is touched 1 second after the file watcher has seen and processed any file event from the OS.

I understand what you are saying but here are the 2 models

**current**
- receive event WRITE tempate/foo.html
- receive event WRITE tempate/foo.html
- receive event WRITE tempate/bar.html
- receive event WRITE tempate/foo.html
- 1 second drain timeout passes
- events are flushed
- handler receives 2 events and uploads
- 1 second notify timeout
- notify happens

**your proposal**
- receive event WRITE tempate/foo.html
- receive event WRITE tempate/foo.html
- receive event WRITE tempate/bar.html
- receive event WRITE tempate/foo.html
- 1 second drain timeout passes
- events are flushed
- handler receives tempate/foo.html event and upload
- notify is touched
- handler receives tempate/bar.html event and upload
- notify is touched

The current situation has the benefit of being closer to when *all* events are processed but may be a bit before the http requests actually finish
Your proposal notifies right after a single http requests is finished but may end up spamming you notifications for every single file event.

**Comment by felipebrahm** [comment]
Thanks for the explanation! But I'd say that from my experience using this, the current implementation is more accurately depicted as this:

**current**

- receive event WRITE template/foo.html
- receive event WRITE assets/big.bundle.js.html
- receive event WRITE template/bar.html
- receive event WRITE template/foo.html
- 1 second drain timeout passes
- events are flushed
- handler receives events and uploads
- file template/bar.html finished uploading
- file template/foo.html finished uploading
- 1 second notify timeout
- **notify happens**
- file assets/big.bundle.js.html finished uploading (**no notification after this!**)

The problem I'm trying to solve is that right now I'm getting notified before all files have finished uploading.

**Comment by tanema** [comment]
Yup I get, that but there is yet another problem that you may still be experiencing with this setup as well. Just because the asset has been uploaded, it does not mean that the asset has been processed and is available yet to your preview so I am not sure this will fix that. We will be looking into it more for sure and I cannot thank you enough for writing the code and really thinking through the problem. 

What service are you using the webhook for?

**Comment by felipebrahm** [comment]
The webhook idea is just to avoid having to rely on touching and watching a file for changes. So it goes like this:

Install a [livereload server implementation](https://github.com/mklabs/tiny-lr) (there are a few, but they all follow the same protocol):
```
npm install --save-dev tiny-lr
```

Create a `livereload-server.js` file:
```javascript
const tinylr = require('tiny-lr');

// standard LiveReload port
const port = 35729;

tinylr().listen(port, () => {
  console.log('Listening on %s', port);
})
```

Then start the live reload server and theme kit (with the `--notify-webhok` addition: https://github.com/felipebrahm/themekit/commit/538de9ddfab3f41b1345082effed633e2f11e62d):
`node livereload-server.js & theme watch --env=development --notify-webhook=\"http://localhost:35729/changed\"`

Then my dev workflow is like this:

- I open Chrome and go to my my-website.myshopify.com
- I enable the [Live Reload Chrome extension ](https://chrome.google.com/webstore/detail/livereload/jnihajbhpnppcggbcgedagnkighmdlei?hl=en)(I could also just inject the JS some other way as part of my local build, but this is the simplest method). Now my page has a websocket to my local live reload server and is listening for changes.
- I change a local file
- themekit is watching for changes, so it uploads the file to Shopify servers
- themekit finishes uploading, so it sends a POST to http://localhost:35729/changed with the details of which file was changed
- my local live reload server receives this POST request and notifies my browser, which is listening via WebSocket
- if it's a CSS change, it tries to apply the change live without reload. if it's any other change, it does a full page reload.

PS: thanks to your explanation above I now know there's a 1s delay before uploading a file (for debouncing purposes), so in my local copy of themekit I modified it to 250ms (for both `pollInterval` and `drainTimeout`) and thus I shaved off 750ms in my live reload :) So another feature request: the `drainTimeout` and `pollInterval` values should probably be configurable via flags.

**Comment by felipebrahm** [comment]
Regarding what you said about "it does not mean that the asset has been processed and is available yet to your preview", at least in my experience this works perfectly fine and the file is available for download as soon as themekit notifies me it has finished uploading it. I've only tried it with files that do not require extra processing though (CSS, not SASS). I'm processing SASS files locally on purpose though, as I just assumed processing them locally and then uploading would make everything faster.

**Comment by tanema** [comment]
Okay I am interested in moving forward with this but with a minor change. Is there a reason you would need both a notify file and the webhook? I am thinking that flag could take a filepath or a URL and just dispatch events depending on what it is.

We will have to test your approach internally to see how it performs with sass and liquid files.

**Comment by felipebrahm** [comment]
> Is there a reason you would need both a notify file and the webhook? I am thinking that flag could take a filepath or a URL and just dispatch events depending on what it is.

I don't think there would be any need to have both available at the same time, so yeah, this makes sense.

**Comment by andyw8** [comment]
@felipebrahm we've released a beta which should address some of your concerns – you can try it with `theme update --version=v1.1.0-beta2`.

**Comment by felipebrahm** [comment]
> @felipebrahm we've released a beta which should address some of your concerns – you can try it with `theme update --version=v1.1.0-beta2`.

Awesome. I just tested the webhook functionality and it works great. Thanks!

**Comment by felipebrahm** [comment]
Fixed by #757 (released on v1.1.0-beta2)
